### DISCUSSION TITLE
tui-input is not reacting to touched

### DISCUSSION THREAD

**Discussion by vytautas-pranskunas-** [taiga-family/taiga-ui#9324] [Q&A] [answered]
I have form with conditional input so lastName becomes required if firstName is not empty. Before submiting form i am doing 
`this.form.markAllAsTouched();`

error message appears but input stays not red.
did not help also 

`
this.form.controls.lastName?.markAsDirty();
            this.form.controls.lastName?.markAsTouched();
`
![Screenshot 2024-10-02 at 21 58 59](https://github.com/user-attachments/assets/5065130d-1282-4f3a-9552-23e2b40524a0)

**Comment by waterplea** [comment, marked as answer]
`markAsTouched` only updates the `touched` state on control with no way of notifying the component. It's one of the longest standing issues for Angular that only got fixed in 18, which we cannot yet use:
https://github.com/angular/angular/issues/10887

We have `tuiMarkControlAsTouchedAndValidate` which you can use to recursively update any `AbstractControl` combination. It calls both `markAsTouched` and `updateValueAndValidity`, which triggers `statusChange` stream. It would also trigger `valueChange` so be careful, maybe you need `distinctUntilChange` somewhere in your streams to not react to that.

P.S. Come to think of it, we can probably typecast the control to `any` and reach for that `events` stream if it's there, that would make it automatic on apps using latest Angular version. I'll add this to the next version 👍
